django/db/models/sql/compiler.py

```diff
@@ -973,6 +973,21 @@ def _get_field_choices():
                     (path, klass_info)
                     for klass_info in klass_info.get('related_klass_infos', [])
                 )
+            for p in ancestor_pointers:
+                yield p.target.name
+
+        def _get_model_ancestor_pointers():
+            """
+            Return a list of OneToOneField relations that are pointers to
+            ancestors of the base 'model'
+            """
+            from django.db.models import OneToOneField
+            ancestors = self.query.get_meta().get_parent_list()
+            cols = [c[0] for c in self.select if isinstance(c[0].target, OneToOneField) and
+                    c[0].target.related_model in ancestors and c[0].target.remote_field.parent_link]
```

Comment:
Only `OneToOneField` can be `parent_link` so this can be reduced to the following at least.

```python
parents = set(self.query.get_meta().get_parent_list())
return [
    c[0] for c in self.select
    if c[0].target.related_model in parents and c[0].target.remote_field.parent_link
]
```